File: Sort/quickSort.py
import math, random, datetime
import arr as randomList
# import insertSort as insert

now = datetime.datetime.now()

'''
基础快速排序
没有重复元素的数组 or 少量重复的数组
'''
def quickSortNoRepeat(sortList) :
    # 基线条件：为空或者只包含一个元素的数组是‘有序’的
    if len(sortList) < 2 : return sortList

    '''
    递归条件
    优化快速排序, 对于近似于有序的数组，使用随机快速排序
    随机生成基准值，最后和数组的第一个值进行数据交换
    '''
    ran = int(math.floor(random.random() * len(sortList)))
    sortList[0], sortList[ran] = sortList[ran], sortList[0]
    pviot = sortList[0]

    # 由所有小于等于基准值的元素组成的子数组
    less = [i for i in sortList[1:] if i <= pviot] # i <= pviot
    # 由所有大于基准值的元素组成的子数组
    greater = [j for j in sortList[1:] if j > pviot]
    return quickSortNoRepeat(less) + [pviot] + quickSortNoRepeat(greater)

quickSortNoRepeat(randomList.generateRandomList(1000, 0, 1000))

now = datetime.datetime.now()

'''
双路快速排序
存在大量的重复元素的数组
'''
def quickSortInTwoWays(sortList, left, right) :
    # 基线条件：为空或者只包含一个元素的数组是‘有序’的
    if right - left < 2 :
        return sortList
    index = partitionInTwoWays(sortList, left, right)

    # 递归条件
    # if index > 1 and index < right - 1 :
    quickSortInTwoWays(sortList, left, index - 1)
    quickSortInTwoWays(sortList, index + 1, right)
    return sortList

def partitionInTwoWays (arr, l, r) :
    # 优化快速排序 => 随机化快速排序, 对于近似于有序的数组，使用随机快速排序
    # 随机生成基准值，最后和数组的第一个值进行数据交换
    ran = int(math.floor(random.random() * (r + l) / 2))
    arr[l], arr[ran] = arr[ran], arr[l]
    pviot = arr[l]

    # arr[l + 1, i) <= pviot; arr(j, r] >= pviot
    i = l
    j = r
    # (i, j)这个区间
    while i <= j :
        while arr[i] < pviot :
            i += 1
        while arr[j] > pviot :
            j -= 1
        if i > j : break
        arr[i], arr[j] = arr[j], arr[i]
        i += 1
        j -= 1
    arr[l], arr[j] = arr[j], arr[l]
    return i

newList = randomList.generateRandomList(300, 0, 10)
left = 0
right = len(newList) - 1

print(quickSortInTwoWays(newList, left, right))
# print(len(quickSortInTwoWays(newList, left, right)))
future = datetime.datetime.now()
# 测试排序速度
print('双路快速排序', future - now)

'''
 三路快速排序
 面向对象封装
'''
lis = [1, 2, 3]
def quickSortInThreeWays (arr, l, r) :
    # 基线条件：为空或者只包含一个元素的数组是‘有序’的
    if r - l < 2 :
        return arr

    # partition
    ran = int(math.floor(random.random() * (r - l) + l))
    arr[l], arr[ran] = arr[ran], arr[l]
    pviot = arr[l]

    lt = l # arr[l + 1...lt]
    gt = r + 1 # arr[gt...r]
    i = l + 1 # arr[lt + 1...i)

    while i < gt :
        if arr[i] < pviot :
            arr[i], arr[lt + 1] = arr[lt + 1], arr[i]
            lt += 1
            i += 1
        elif arr[i] > pviot :
            arr[i], arr[gt - 1] = arr[gt - 1], arr[i]
            gt -= 1
        else :
            i += 1
    arr[l], arr[lt] = arr[lt], arr[l]

    # 递归条件
    quickSortInThreeWays(arr, l, lt - 1)
    quickSortInThreeWays(arr, gt, r)
    return arr

newList = randomList.generateRandomList(300, 0, 10)
left = 0
right = len(newList) - 1
print(quickSortInThreeWays(newList, left, right))
print(len(quickSortInThreeWays(newList, left, right)))

ai = datetime.datetime.now()
# 测试排序速度
print('三路快速排序', ai - future)

'''
 双路快速排序
 面向对象封装
'''
class quick_sort():
    def _partition(self, alist, p, r):
        i = p - 1
        x = alist[r]
        for j in range(p, r):
            if alist[j] <= x:
                i += 1
                alist[i], alist[j] = alist[j], alist[i]
        alist[i + 1], alist[r] = alist[r], alist[i+1]
        return i + 1

    def _quicksort(self, alist, p, r):
        if p<r:
            q = self._partition(alist, p, r)
            self._quicksort(alist, p, q - 1)
            self._quicksort(alist, q + 1, r)

    def __call__(self, sort_list):
        self._quicksort(sort_list, 0, len(sort_list) - 1)
        return sort_list

print(quick_sort)

File: Sort/quickSort.js
function quickSort (arr, left, right) {
  if (right - left < 2) {
    return arr;
  }

  let index = partition(arr, left, right)

  quickSort(arr, left, index - 1)
  quickSort(arr, index, right)
  return arr
}

function partition (arr, left, right) {
  // 添加随机性 Math.random，优化近似于有序的数组
  let pivot = arr[Math.floor(Math.random() * (right - left + 1) + left)];
  let i = left;
  let j = right;

  while (i <= j) {
    while (arr[i] < pivot) {
      i++;
    }

    while (arr[j] > pivot) {
      j--;
    }

    if (i > j) break

    [arr[i], arr[j]] = [arr[j], arr[i]]
    i++;
    j--;
  }

  return i
}

let arr = [6, 2, 5, 1, 8, 2, 2, 10, 2, 6, 5, 9, 2, 7, 7, 9, 7, 0, 4, 5, 10, 1, 3, 3, 2, 5, 5, 9, 0, 4, 7, 2, 8, 9, 9, 3, 6, 9, 8, 10, 2,
 0, 2, 7, 10, 2, 3, 9, 4, 6, 7, 4, 7, 2, 8, 8, 4, 0, 3, 5, 2, 2, 9, 2, 5, 1, 4, 3, 10, 4, 2, 9, 4, 10, 4, 0, 0, 2, 1, 0, 5, 10
, 9, 3, 2, 0, 8, 7, 5, 9, 5, 8, 2, 10, 3, 1, 5, 10, 4, 8, 9, 7, 10, 10, 10, 10, 7, 7, 3, 6, 4, 6, 6, 2, 4, 4, 4, 4, 2, 10, 5,
6, 9, 3, 2, 1, 0, 9, 10, 10, 2, 6, 8, 7, 0, 0, 8, 9, 4, 6, 1, 5, 0, 10, 9, 5, 0, 0, 1, 7, 10, 1, 10, 7, 8, 1, 3, 8, 10, 1, 9,
3, 10, 4, 0, 2, 4, 2, 9, 10, 0, 8, 0, 2, 10, 10, 2, 5, 5, 6, 1, 1, 9, 3, 3, 5, 6, 6, 9, 1, 1, 4, 7, 5, 9, 2, 2, 6, 7, 1]
quickSort(arr, 0, arr.length - 1)


/**
 * 三路快排
 * @param {*} arr 
 * @param {*} l 
 * @param {*} r 
 */
function quickSortIn3Ways (arr, l, r) {
  if (r - l < 2) return arr

  let ran = Math.floor(Math.random() * (r - l + 1) + l)
  swapArr(arr, l, ran)
  // [arr[l], arr[ran]] = [arr[ran], arr[l]]
  let pvoit = arr[l]

  let lt = l // arr[l+1, lt]
  let gt = r + 1 // arr[gt, r]
  let i = l + 1 // arr[lt + 1, i)

  while (i < gt) {
    if (arr[i] < pvoit) {
      [arr[i], arr[lt + 1]] = [arr[lt + 1], arr[i]]
      lt++
      i++
    } else if (arr[i] > pvoit) {
      [arr[i], arr[gt - 1]] = [arr[gt - 1], arr[i]]
      gt--
    } else {
      i++
    }
  }

  [arr[l], arr[lt]] = [arr[lt], arr[l]]

  quickSortIn3Ways(arr, l, lt - 1)
  quickSortIn3Ways(arr, gt, r)
  return arr
}

function swapArr (arr, a, b) {
  [arr[a], arr[b]] = [arr[b], arr[a]]
}

let arr = [6, 2, 5, 1, 8, 2, 2, 10, 2, 6, 5, 9, 2, 7, 7, 9, 7, 0, 4, 5, 10, 1, 3, 3, 2, 5, 5, 9, 0, 4, 7, 2, 8, 9, 9, 3, 6, 9, 8, 10, 2,
  0, 2, 7, 10, 2, 3, 9, 4, 6, 7, 4, 7, 2, 8, 8, 4, 0, 3, 5, 2, 2, 9, 2, 5, 1, 4, 3, 10, 4, 2, 9, 4, 10, 4, 0, 0, 2, 1, 0, 5, 10
 , 9, 3, 2, 0, 8, 7, 5, 9, 5, 8, 2, 10, 3, 1, 5, 10, 4, 8, 9, 7, 10, 10, 10, 10, 7, 7, 3, 6, 4, 6, 6, 2, 4, 4, 4, 4, 2, 10, 5,
 6, 9, 3, 2, 1, 0, 9, 10, 10, 2, 6, 8, 7, 0, 0, 8, 9, 4, 6, 1, 5, 0, 10, 9, 5, 0, 0, 1, 7, 10, 1, 10, 7, 8, 1, 3, 8, 10, 1, 9,
 3, 10, 4, 0, 2, 4, 2, 9, 10, 0, 8, 0, 2, 10, 10, 2, 5, 5, 6, 1, 1, 9, 3, 3, 5, 6, 6, 9, 1, 1, 4, 7, 5, 9, 2, 2, 6, 7, 1]
quickSortIn3Ways(arr, 0, arr.length - 1)
